File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.js
/**
 * @param {number[]} nums
 * @param {number} limit
 * @return {number}
 */
var longestSubarray = function(nums, limit) {
  const qmax = [];
  const qmin = [];
  let [l1, l2, r1, r2, i] = [0, 0, -1, -1, 0];
  
  for (let num of nums) {
    while (l1 <= r1 && qmax[r1] < num) r1--;
    while (l2 <= r2 && qmin[r2] > num) r2--;
    
    qmax[++r1] = num;
    qmin[++r2] = num;
    
    if (l1 <= r1 && l2 <= r2 && qmax[l1] - qmin[l2] > limit) {
      if (qmax[l1] === nums[i]) l1++;
      if (qmin[l2] === nums[i]) l2++;
      i++;
    }
  }
  
  return nums.length - i;
};

File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.py
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        qmax, qmin, i = deque(), deque(), 0
        for num in nums:
            while qmax and qmax[-1] < num:
                qmax.pop()
            while qmin and qmin[-1] > num:
                qmin.pop()
            qmax.append(num)
            qmin.append(num)
            if qmax[0] - qmin[0] > limit:
                if qmax[0] == nums[i]:
                    qmax.popleft()
                if qmin[0] == nums[i]:
                    qmin.popleft()
                i += 1
        return len(nums) - i;